framework/behaviors/AttributeTypecastBehavior.php

```diff
@@ -244,13 +248,18 @@ public function typecastAttributes($attributeNames = null)
 
     /**
      * Casts the given value to the specified type.
+     *
      * @param mixed $value value to be type-casted.
-     * @param string|callable $type type name or typecast callable.
+     * @param string|callable $type type name or typecast callback: `function (mixed $value): mixed`
+     * 
      * @return mixed typecast result.
      */
     protected function typecastValue($value, $type)
     {
-        if (is_scalar($type)) {
+        if (
+            in_array($type, [self::TYPE_INTEGER, self::TYPE_FLOAT, self::TYPE_BOOLEAN, self::TYPE_STRING], true)
+            || !is_callable($type)
```

Comment:
I think this condition (or the previous one deleted) could be negated and `return call_user_func($type, $value);` could be done early. That would simplify the code a bit.

Reply:
`is_callable` is now called on each type-casting. That doesn't look great performance-wise.

Reply:
@samdark nope, if `in_array()` return true, then `is_callable()`

Reply:
If `is_callable()` return `false` for example for array, `throw new InvalidArgumentException("Unsupported type '$type'")` will fail because you're trying to cast array to string.